models/demos/mistral7b/tt/mistral_attention.py

```diff
@@ -0,0 +1,348 @@
+
+        self.n_local_heads = self.n_heads // self.num_devices
+        self.n_local_kv_heads = self.n_kv_heads // self.num_devices
+
+        self.dtype = dtype
+
+        layer_name = f"layers.{layer_num}.attention"
+        cache_name = lambda name: weight_cache_path / (f"{layer_name}.{name}")
+
+        wq_str = f"{layer_name}.wq.weight"
+        wk_str = f"{layer_name}.wk.weight"
+        wv_str = f"{layer_name}.wv.weight"
+        wo_str = f"{layer_name}.wo.weight"
+
+        # when splitting the devices, we need to make sure that the number of heads is divisible by the number of devices
+        assert self.n_heads % self.num_devices == 0
+        assert self.n_kv_heads % self.num_devices == 0
+
+        self.wqkv_list = []
+        self.wo_list = []
+        self.layer_past_list = []
+
+        self.kernel_config = ttnn.WormholeComputeKernelConfig(
+            math_fidelity=ttnn.MathFidelity.HiFi4,
+            math_approx_mode=False,
+            fp32_dest_acc_en=True,
+            packer_l1_acc=True,
+        )
+
+        for i in range(self.num_devices):
+            wqkv = ttnn.as_tensor(
+                torch.concat(
+                    [
+                        torch.transpose(
+                            torch.chunk(self.state_dict[wq_str], self.num_devices)[i],
+                            -2,
+                            -1,
+                        ),
+                        torch.transpose(
+                            torch.chunk(self.state_dict[wk_str], self.num_devices)[i],
+                            -2,
+                            -1,
+                        ),
+                        torch.transpose(
+                            torch.chunk(self.state_dict[wv_str], self.num_devices)[i],
+                            -2,
+                            -1,
+                        ),
+                    ],
+                    dim=-1,
+                ),
+                device=self.devices[i],
+                dtype=self.dtype,
+                memory_config=ttnn.DRAM_MEMORY_CONFIG,
+                layout=ttnn.TILE_LAYOUT,
+                cache_file_name=cache_name("wqkv"),
+            )
+
+            wo = ttnn.as_tensor(
+                torch.transpose(
+                    torch.chunk(self.state_dict[wo_str], self.num_devices, dim=-1)[i],
+                    -2,
+                    -1,
+                ),
+                device=self.devices[i],
+                memory_config=ttnn.DRAM_MEMORY_CONFIG,
+                dtype=self.dtype,
+                layout=ttnn.TILE_LAYOUT,
+                cache_file_name=cache_name("wo"),
+            )
+
+            cache_k = torch.zeros(
+                (
+                    self.max_batch_size,
+                    self.n_kv_heads // self.num_devices,
+                    # self.sliding_window,
+                    512,  # TODO Update the initial cache size when scaling up (Target = window size == 4096)
```

Comment:
Is 512 the seq_len? Should this be something we pass in to the model?

Reply:
Yes. This should be equal to `self.sliding_window` == 4096, but that would be too large to fit on DRAM for a single chip.

I could add this as a parameter to model_config instead, to keep the attn module cleaner.

Reply:
I moved this to model_config.